menas/src/main/scala/za/co/absa/enceladus/menas/services/DatasetService.scala

```diff
@@ -102,4 +113,177 @@ class DatasetService @Autowired() (datasetMongoRepository: DatasetMongoRepositor
     }
   }
 
+  override def importItem(item: Dataset, username: String): Future[Option[Dataset]] = {
+    getLatestVersionValue(item.name).flatMap {
+      case Some(version) => update(username, item.copy(version = version))
+      case None => super.create(item.copy(version = 1), username)
+    }
+  }
+
+  override def validateSingleImport(item: Dataset, metadata: Map[String, String]): Future[Validation] = {
+    val confRulesWithConnectedEntities = item.conformance.filter(_.hasConnectedEntities)
+    val maybeSchema = datasetMongoRepository.getConnectedSchema(item.schemaName, item.schemaVersion)
+
+    for {
+      validationBase <- super.validateSingleImport(item, metadata)
+      validationSchema <- validateSchema(item.schemaName, item.schemaVersion, maybeSchema)
+      validationConnectedEntities <- validateConnectedEntitiesExistence(confRulesWithConnectedEntities)
+      validationConformanceRules <- validateConformanceRules(item.conformance, maybeSchema)
+    } yield validationBase.merge(validationSchema).merge(validationConnectedEntities).merge(validationConformanceRules)
+  }
+
+  private def validateConnectedEntitiesExistence(confRulesWithConnectedEntities: List[ConformanceRule]): Future[Validation] = {
+    def standardizedErrMessage(ce: ConnectedEntity) = s"Connected ${ce.kind} ${ce.name} v${ce.version} could not be found"
+
+    confRulesWithConnectedEntities.foldLeft(Future(Validation())) { (acc, cr) =>
+      cr.connectedEntities.foldLeft(acc) { (validation, connectedEntity)  =>
+        connectedEntity match {
+          case mt: ConnectedMappingTable =>
+            for {
+              instance <- datasetMongoRepository.getConnectedMappingTable(mt.name, mt.version)
+              accValidations <- validation
+            } yield instance
+              .map { _ => accValidations.withError(s"item.${mt.kind}", standardizedErrMessage(mt))}
+              .getOrElse(accValidations)
+        }
+      }
+
+      val connectedEntities = cr.connectedEntities.map {
+        case mt: ConnectedMappingTable => (mt, datasetMongoRepository.getConnectedMappingTable(mt.name, mt.version))
+      }
+
+      connectedEntities.foldLeft(acc) { case (v, (entityDef, entityDBInstance)) =>
+        for {
+          instance <- entityDBInstance
+          accValidations <- v
+        } yield accValidations.withErrorIf(instance.isEmpty, s"item.${entityDef.kind}", standardizedErrMessage(entityDef))
+      }
```

Comment:
I took the liberty to suggest an alternate solution for the whole function. Has the advantage, that if there are more of the same connected entities, there will be only one validation for all of them.
```suggestion
    val allConnectedEntites: Set[ConnectedEntity] = confRulesWithConnectedEntities.foldLeft(Set.empty[ConnectedEntity]) { (acc, cr) =>
      acc ++ cr.connectedEntities.toSet
    }
    
    allConnectedEntites.foldLeft(Future(Validation())) { (accValidations, entityDef) =>
      entityDef match {
        case mt: ConnectedMappingTable =>
          val entityDBInstance = datasetMongoRepository.getConnectedMappingTable(mt.name, mt.version)
          for {
            instance <- entityDBInstance
            validations <- accValidations
          } yield validations.withErrorIf(instance.isEmpty, s"item.${entityDef.kind}", standardizedErrMessage(entityDef))
      }
    }
```

Reply:
Fixed